File: python/test/test_zmq_communication.py
import numpy as np
import state_representation as sr
import threading
import time
import unittest
import zmq
from network_interfaces.zmq import network
from numpy.testing import assert_array_almost_equal


class TestZMQNetworkInterface(unittest.TestCase):
    robot_state = None
    robot_joint_state = None
    robot_jacobian = None
    robot_mass = None
    control_command = None
    control_type = None
    context = None
    received_state = None
    received_command = None

    @classmethod
    def setUpClass(cls):
        cls.robot_state = sr.CartesianState().Random("ee", "robot")
        cls.robot_joint_state = sr.JointState().Random("robot", 3)
        cls.robot_jacobian = sr.Jacobian().Random("robot", 3, "frame")
        cls.robot_mass = sr.Parameter("mass", np.random.rand(3, 3), sr.ParameterType.MATRIX)
        cls.control_command = sr.JointState().Random("robot", 3)
        cls.control_type = [1, 2, 3]
        cls.context = zmq.Context()

    def assert_state_equal(self, state1, state2):
        self.assertTrue(state1.is_compatible(state2))
        assert_array_almost_equal(state1.data(), state2.data())

    def robot(self):
        command_subscriber, state_publisher = network.configure_sockets(self.context, "127.0.0.1:1702",
                                                                        "127.0.0.1:1701")

        state = network.StateMessage(self.robot_state, self.robot_joint_state, self.robot_jacobian, self.robot_mass)
        command = []
        for i in range(200):
            network.send_state(state, state_publisher)
            command = network.receive_command(command_subscriber)
            time.sleep(0.01)
        command_subscriber.close()
        state_publisher.close()
        self.received_command = command

    def control(self):
        state_subscriber, command_publisher = network.configure_sockets(self.context, "127.0.0.1:1701",
                                                                        "127.0.0.1:1702")

        command = network.CommandMessage(self.control_type, self.control_command)
        state = []
        for i in range(200):
            network.send_command(command, command_publisher)
            state = network.receive_state(state_subscriber)
            time.sleep(0.01)
        state_subscriber.close()
        command_publisher.close()
        self.received_state = state

    def test_communication(self):
        robot = threading.Thread(target=self.robot)
        control = threading.Thread(target=self.control)
        robot.start()
        control.start()
        robot.join()
        control.join()

        [self.assertEqual(self.received_command.control_type[i], self.control_type[i]) for i in
         range(len(self.control_type))]
        self.assert_state_equal(self.received_command.joint_state, self.control_command)

        self.assert_state_equal(self.received_state.ee_state, self.robot_state)
        self.assert_state_equal(self.received_state.joint_state, self.robot_joint_state)
        self.assert_state_equal(self.received_state.jacobian, self.robot_jacobian)
        self.assertTrue(self.received_state.mass.is_compatible(self.robot_mass))
        assert_array_almost_equal(self.received_state.mass.get_value(), self.robot_mass.get_value())

    def test_encode_command(self):
        command = network.CommandMessage([], sr.JointState())
        network.encode_command(command)

        command = network.CommandMessage([], sr.JointState().Random("test", 3))
        with self.assertRaises(ValueError):
            network.encode_command(command)

        command.control_type = [1]
        encoded = network.encode_command(command)
        decoded = network.decode_command(encoded)
        self.assertEqual(len(decoded.control_type), 3)
        assert_array_almost_equal(decoded.control_type, [1, 1, 1])

        command.control_type = [1, 2, 6]
        with self.assertRaises(ValueError):
            network.encode_command(command)

        command.control_type = [1, 2, 2, 1]
        with self.assertRaises(ValueError):
            network.encode_command(command)

File: cpp/test/test_zmq_communication.cpp
#include <gtest/gtest.h>
#include <thread>

#include "network_interfaces/zmq/network.h"
#include "network_interfaces/zmq/exceptions/InvalidControlTypeVectorException.h"
#include "network_interfaces/zmq/exceptions/UnknownControlTypeException.h"

static void
expect_joint_state_equal(const state_representation::JointState& js1, const state_representation::JointState& js2) {
  EXPECT_TRUE(js1.data().isApprox(js2.data()));
  EXPECT_FALSE(js1.is_incompatible(js2));
}

static void expect_cart_state_equal(
    const state_representation::CartesianState& cs1, const state_representation::CartesianState& cs2
) {
  EXPECT_TRUE(cs1.data().isApprox(cs2.data()));
  EXPECT_FALSE(cs1.is_incompatible(cs2));
}

class TestNetworkInterface : public ::testing::Test {
public:
  TestNetworkInterface() {
    robot_state = state_representation::CartesianState::Random("ee", "robot");
    robot_joint_state = state_representation::JointState::Random("robot", 3);
    robot_jacobian = state_representation::Jacobian::Random("robot", 3, "frame");
    robot_mass.set_value(Eigen::MatrixXd::Random(3, 3));
    control_joint_state = state_representation::JointState::Random("robot", 3);
    control_type = std::vector<int>{1, 2, 3};
  }

  state_representation::CartesianState robot_state;
  state_representation::JointState robot_joint_state;
  state_representation::Jacobian robot_jacobian;
  state_representation::Parameter<Eigen::MatrixXd>
      robot_mass = state_representation::Parameter<Eigen::MatrixXd>("robot");
  state_representation::JointState control_joint_state;
  std::vector<int> control_type;

  std::thread spawn_robot() {
    return std::thread([this] { this->robot(); });
  }

  void robot() const {
    zmq::context_t context = zmq::context_t(1);
    zmq::socket_t state_publisher, command_subscriber;
    network_interfaces::zmq::configure_sockets(
        context, command_subscriber, "127.0.0.1:1702", state_publisher, "127.0.0.1:1701"
    );

    network_interfaces::zmq::CommandMessage command;
    network_interfaces::zmq::StateMessage state;
    state.ee_state = robot_state;
    state.joint_state = robot_joint_state;
    state.jacobian = robot_jacobian;
    state.mass = robot_mass;
    for (std::size_t i = 0; i < 100; ++i) {
      network_interfaces::zmq::send(state, state_publisher);
      network_interfaces::zmq::receive(command, command_subscriber);
      usleep(10000);
    }
    expect_joint_state_equal(command.joint_state, control_joint_state);
    for (std::size_t i = 0; i < control_type.size(); ++i) {
      EXPECT_EQ(control_type.at(i), command.control_type.at(i));
    }
  }

  std::thread spawn_control() {
    return std::thread([this] { this->control(); });
  }

  void control() const {
    zmq::context_t context = zmq::context_t(1);
    zmq::socket_t command_publisher, state_subscriber;
    network_interfaces::zmq::configure_sockets(
        context, state_subscriber, "127.0.0.1:1701", command_publisher, "127.0.0.1:1702"
    );

    network_interfaces::zmq::StateMessage state;
    network_interfaces::zmq::CommandMessage command;
    command.joint_state = control_joint_state;
    command.control_type = control_type;
    for (std::size_t i = 0; i < 100; ++i) {
      network_interfaces::zmq::send(command, command_publisher);
      network_interfaces::zmq::receive(state, state_subscriber);
      usleep(10000);
    }
    expect_cart_state_equal(state.ee_state, robot_state);
    expect_joint_state_equal(state.joint_state, robot_joint_state);
    EXPECT_TRUE(state.jacobian.data().isApprox(robot_jacobian.data()));
    EXPECT_FALSE(state.jacobian.is_incompatible(robot_jacobian));
    EXPECT_TRUE(state.mass.get_value().isApprox(robot_mass.get_value()));
    EXPECT_FALSE(state.mass.is_incompatible(robot_mass));
  }
};

TEST_F(TestNetworkInterface, TestEncodeCommand) {
  network_interfaces::zmq::CommandMessage command;
  command.joint_state = state_representation::JointState::Random("test", 3);
  command.control_type = std::vector<int>{};
  EXPECT_THROW(network_interfaces::zmq::encode_command(command),
               network_interfaces::zmq::exceptions::InvalidControlTypeVectorException);

  command.control_type = std::vector<int>{1};
  network_interfaces::zmq::encode_command(command);
  ASSERT_EQ(command.control_type.size(), 3);
  for (std::size_t i = 0; i < 3; ++i) {
    EXPECT_EQ(command.control_type.at(i), 1);
  }

  command.control_type = std::vector<int>{1, 2, 6};
  EXPECT_THROW(network_interfaces::zmq::encode_command(command),
               network_interfaces::zmq::exceptions::UnknownControlTypeException);

  command.control_type = std::vector<int>{1, 2, 2, 1};
  EXPECT_THROW(network_interfaces::zmq::encode_command(command),
               network_interfaces::zmq::exceptions::InvalidControlTypeVectorException);
}

TEST_F(TestNetworkInterface, TestEmptyEncodeCommand) {
  network_interfaces::zmq::CommandMessage command;
  EXPECT_NO_THROW(network_interfaces::zmq::encode_command(command));

  command.control_type = std::vector<int>{1};
  EXPECT_THROW(network_interfaces::zmq::encode_command(command),
               network_interfaces::zmq::exceptions::InvalidControlTypeVectorException);
}

TEST_F(TestNetworkInterface, TestCommunication) {
  std::thread robot = spawn_robot();
  std::thread control = spawn_control();

  robot.join();
  control.join();
}

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}